source/elements/oneMKL/source/domains/spblas/operations/spmv.rst

```diff
@@ -0,0 +1,378 @@
+       | cuSPARSE: ``CUSPARSE_SPMV_COO_ALG2``
+       | rocSPARSE: ``rocsparse_spmv_alg_coo_atomic``
+   * - ``csr_alg1``
+     - Default algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_CSR_ALG1``
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_adaptive``
+   * - ``csr_alg2``
+     - Deterministic algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: ``CUSPARSE_SPMV_CSR_ALG2``
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_stream``
+   * - ``csr_alg3``
+     - LRB variant of the algorithm for CSR format.
+     - | MKL: N/A
+       | cuSPARSE: N/A
+       | rocSPARSE: ``rocsparse_spmv_alg_csr_lrb``
+
+.. _onemkl_sparse_spmv:
+
+spmv
+----
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       void spmv_buffer_size(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           std::size_t                                &temp_buffer_size,
+       );
+
+       void spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           sycl::buffer<std::uint8_t, 1>              workspace,
+       );
+
+       sycl::event spmv_optimize(
+           sycl::queue                                &queue,
+           oneapi::mkl::transpose                     opA,
+           void*                                      alpha,
+           oneapi::mkl::sparse::matrix_view           A_view,
+           oneapi::mkl::sparse::matrix_handle_t       A_handle,
+           oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+           void*                                      beta,
+           oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+           oneapi::mkl::sparse::spmv_alg              alg,
+           oneapi::mkl::sparse::spmv_descr_t          spmv_descr,
+           void*                                      workspace,
```

Comment:
There is a `workspace` arg in `spmv_optimize()` but not in `spmv()` API. Maybe we should mention that `spmv_descr` object stores the `workspace` pointer inside for later use.

We also need to make a decision here for the Specifcation regarding reusability of `spmv_descr` and/or `workspace` across different input combinations (viz., `opA`, `A_view`, and even `A_handle`). Even if we allow reusability, we cannot allow interleaving `spmv_optimize`/`spmv` API calls between different input combinations. `spmv_buffer_size` calls can be interleaved so that the allocated workspace size can be `std::max` over returned sizes.

For reusing `workspace`, the following is the most achievable level of flexibility:
```cpp
init_spmv_descr(queue, spmv_descr);

/* At least one of {opA_1, opA_2}, {viewA_1, viewA_2}, {handleA_1, handleA_2} pairs are different */
spmv_buffer_size(..., opA_1, viewA_1, handleA_1, ..., buffersize1);
spmv_buffer_size(..., opA_2, viewA_2, handleA_2, ..., buffersize2);
void *workspace = malloc_device(std::max(buffersize1, buffersize2), ...);

spmv_optimize(..., opA_1, viewA_1, handleA_1, ..., spmv_descr, workspace, ...);
spmv(..., opA_1, viewA_1, handleA_1, ,..., spmv_descr, ...); // Call as many times as you want

spmv_optimize(..., opA_2, viewA_2, handleA_2, ..., spmv_descr, workspace, ...);
spmv(..., opA_2, viewA_2, handleA_2, ..., spmv_descr, ...); // Call as many times as you want

release_spmv_descr(...);
```

Reply:
Calling `spmv_optimize` resets or overwrites the `workspace` data every time.

Reply:
I have made some clarifications on the `workspace` in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/3ab3c5e011abe78133fc917617f91325535f64ce and on the usage of `*_optimize` functions in https://github.com/oneapi-src/oneAPI-spec/pull/522/commits/15ec0e501cb06a4a957376dd9525cb1babe8e644. I think that covers what you are describing.

Reply:
> Maybe we should mention that spmv_descr object stores the workspace pointer inside for later use.

Are we mentioning that somewhere in the spec, either around the description of `spmv` or `spmv_descr`?

Reply:
Yes, in the commit above I added the following:

> Calling ``spmv_optimize`` on the same descriptor can reset
     some of the descriptor's data such as the ``workspace``.